src/detect-modbus.c

```diff
@@ -572,6 +176,72 @@ void DetectModbusRegister(void)
 #ifdef UNITTESTS /* UNITTESTS */
 #include "util-unittest.h"
 
+/** Convert rust structure to C for regression tests.
+ *
+ * Note: Newly allocated `DetectModbus` structure must be freed.
+ *
+ * TODO: remove this after regression testing commit.
+ */
+static DetectModbusValue *DetectModbusValueRustToC(uint16_t min, uint16_t max)
+{
+    DetectModbusValue *value = SCMalloc(sizeof(*value));
+    FAIL_IF_NULL(value);
+
+    value->min = min;
+    value->max = max;
+
+    if (min == max) {
+        value->mode = DETECT_MODBUS_EQ;
+    } else if (min == 0) {
+        value->mode = DETECT_MODBUS_LT;
+    } else if (max == UINT16_MAX) {
+        value->mode = DETECT_MODBUS_GT;
+    } else {
+        value->mode = DETECT_MODBUS_RA;
+    }
+
+    return value;
+}
+
+static DetectModbus *DetectModbusRustToC(DetectModbusRust *ctx)
+{
+    DetectModbus *modbus = SCMalloc(sizeof(*modbus));
+    FAIL_IF_NULL(modbus);
+
+    modbus->category = rs_modbus_get_category(ctx);
+    modbus->function = rs_modbus_get_function(ctx);
+    modbus->subfunction = rs_modbus_get_subfunction(ctx);
+    modbus->has_subfunction = rs_modbus_get_has_subfunction(ctx);
+    modbus->type = rs_modbus_get_access_type(ctx);
+
+    modbus->unit_id = DetectModbusValueRustToC(
+            rs_modbus_get_unit_id_min(ctx), rs_modbus_get_unit_id_max(ctx));
+
+    modbus->address = DetectModbusValueRustToC(
+            rs_modbus_get_address_min(ctx), rs_modbus_get_address_max(ctx));
+
+    modbus->data =
+            DetectModbusValueRustToC(rs_modbus_get_data_min(ctx), rs_modbus_get_data_max(ctx));
+
+    return modbus;
+}
+
+static void DetectModbusCFree(DetectModbus *modbus)
+{
+    if (modbus) {
+        if (modbus->unit_id)
+            SCFree(modbus->unit_id);
+
+        if (modbus->address)
+            SCFree(modbus->address);
+
+        if (modbus->data)
+            SCFree(modbus->data);
+
+        SCFree(modbus);
+    }
+}
+
 /** \test Signature containing a function. */
 static int DetectModbusTest01(void)
```

Comment:
@catenacyber We tried to change the unit tests as little as possible in this commit to test the rust implementation using C tests before moving them to Rust. There are some minor changes, as noted in the description of the PR. More tests are also in `detect-engine-modbus.c`. Unfortunately, we could not do the same with the `app-layer-modbus.c` tests because this would be significantly more effort.

Reply:
Thank you for this.

How did you change the test `DetectModbusTest04` ?
Why was it not possible for app-layer-modbus.c ?
Reading `ModbusParserTest01`, was it because of accessing internal transaction fields ? like ` FAIL_IF_NOT(tx->read.address == 0x7890);` ?

Reply:
> How did you change the test `DetectModbusTest04` ?

The actual diff is [here](https://github.com/OISF/suricata/pull/5810/commits/a51e82e508c8df47b7b3959e9e2046997e7a1919#diff-a8bc8a6c2bef032dcc859072e8031de5cd492227114fb74cd9079b569edf9400R357).
Instead of comparing to `~MODBUS_CAT_PUBLIC_ASSIGNED` we are checking all the flags individually.
```rust

    FAIL_IF(modbus->category & MODBUS_CAT_PUBLIC_ASSIGNED);
    FAIL_IF_NOT(modbus->category & MODBUS_CAT_PUBLIC_UNASSIGNED);
    FAIL_IF_NOT(modbus->category & MODBUS_CAT_USER_DEFINED);
    FAIL_IF_NOT(modbus->category & MODBUS_CAT_RESERVED); 
```
> Why was it not possible for app-layer-modbus.c ?

I'll see what I can do :).

> Reading `ModbusParserTest01`, was it because of accessing internal transaction fields ? like ` FAIL_IF_NOT(tx->read.address == 0x7890);` ?

Yes, I will have to create temporary ffi accessors for the test cases to work like in `detect-modbus.c`. I'll see how much effort this is.

Reply:
Looks good for `DetectModbusTest04 `
Thanks for the excellent explanation.

Temporary ffi accessors seem indeed too much work...
Is there a better way @jasonish ?
Is it possible to share the structure between Rust and C ?

In the meantime, how did you translate this specific `ModbusParserTest01` ? (I do not see such a name in the rust tests)

Reply:
I think it will be similar effort either way. The advantage of writing accessors is that we don't have to refactor the C tests to make them compatible with the rust structs (the diff remains as small as possible).

That test was moved to [rust/src/modbus/modbus.rs](https://github.com/OISF/suricata/pull/5810/files?file-filters%5B%5D=.am&file-filters%5B%5D=.c&file-filters%5B%5D=.h&file-filters%5B%5D=.rs#diff-57a6cf267e4fb39aa835b56cc817f0c7dccbafc37de20f4c90eaf8eb216a5958R683). It does the same function, address, qt check, calls parse again for the response, adds a couple more checks and also checks transaction length.

All tests in that file should be the same order as the C tests.
- `ModbusParserTest12` wasn't added because it calls `UTHBuildPacket`whereas we are testing the parse function.
- `ModbusParserTest18` wasn't added because there is no stream depth in the new module. Not sure how this is handled by `RustParser`.

Reply:
Ok, so `read_coils` is the new `ModbusParserTest01` right ?

For `ModbusParserTest12` what is the replacement ? A suricata-verify test to check that we get an alert ?

`ModbusParserTest18` may need some more investigation... The `reassembly_depth` seems generic about TCP...

Reply:
All of the `ModbusParserTest*` were re-introduced in #5849 [app-layer-modbus.c](https://github.com/OISF/suricata/blob/507f811f9d8be83638e7c2e4793fc65f69929e01/src/app-layer-modbus.c#L267).

I am not sure how `reassembly_depth` works with rust parsers and started looking at `stream-tcp-reassmble.c`. If you have any relevant documentation that would be helpful.

Reply:
You can take a look at https://suricata.readthedocs.io/en/suricata-6.0.0/configuration/suricata-yaml.html?highlight=reassembly%20depth#stream-engine

It should work the same for C and Rust parsers (as it is in TCP handling, not in the app-layer handling)